automata/fa/fa.py

```diff
@@ -2,11 +2,245 @@
 """Classes and methods for working with all finite automata."""
 
 import abc
+import itertools
+import os
+import pathlib
+import typing
+import uuid
+from collections import defaultdict
+
+import graphviz
+from coloraide import Color
 
 from automata.base.automaton import Automaton
 
 
 class FA(Automaton, metaclass=abc.ABCMeta):
     """An abstract base class for finite automata."""
 
-    pass
+    @staticmethod
+    def get_state_name(state_data):
+        """
+        Get an string representation of a state. This is used for displaying and
+        uses `str` for any unsupported python data types.
+        """
+        if isinstance(state_data, str):
+            if state_data == "":
+                return "λ"
+
+            return state_data
+
+        if isinstance(state_data, (set, frozenset, list, tuple)):
+            inner = ", ".join(FA.get_state_name(sub_data) for sub_data in state_data)
+            if isinstance(state_data, (set, frozenset)):
+                return '{' + inner + '}'
+
+            if isinstance(state_data, tuple):
+                return '(' + inner + ')'
+
+            if isinstance(state_data, list):
+                return '[' + inner + ']'
+
+        return str(state_data)
+
+    @abc.abstractmethod
+    def iter_states(self):
+        """Iterate over all states in the automaton."""
+
+    @abc.abstractmethod
+    def iter_transitions(self):
+        """
+        Iterate over all transitions in the automaton. Each transition is a tuple
+        of the form (from_state, to_state, symbol)
+        """
+
+    @abc.abstractmethod
+    def is_accepting(self, state):
+        """Check if a state is an accepting state."""
+
+    @abc.abstractmethod
+    def is_initial(self, state):
+        """Check if a state is an initial state."""
+
+    def show(
```

Comment:
@khoda81 Why is this method called `show` and not `show_diagram`? I would much prefer the latter, and not sure why `show_diagram` was removed. cc @eliotwrobson

Reply:
That was a temporary change to slowly move the code from `show_diagram` to this new method. I will change the name in the back to `show_diagram` in the next commit.